docs/settings.md
# Configuration fields
(This page was autogenerated, do not edit manually)
### init
Create a new tasks.py file in the current directory  
(no env var)  

### tags
Only show tasks matching any of these tags  
TASKCLI_CFG_TAGS  

### group_order
Regex re.match patterns of the order in which top-level groups are shown. Any top-level group the name of which matches any of the patterns will be listed first.  
TASKCLI_CFG_GROUP_ORDER  

### search
Only show tasks whose name or description is matching this python regex seach pattern.  
TASKCLI_CFG_SEARCH  

### parent
Whether to also include tasks from the closest parent's directory's tasks.py  
TASKCLI_CFG_PARENT  

### color
Whether or not the output should be colored  
(no env var)  

### show_hidden
Show all tasks and groups, even the hidden ones.  
TASKCLI_CFG_SHOW_HIDDEN  

### run_show_location
When true, each call to tt.run() also logs the code location from which it was called  
TASKCLI_CFG_RUN_SHOW_LOCATION  

### no_go_task
Disable automatic inclusion of tasks from 'task' binary. Note, for this automaic inclusion to work, TASKCLI_GOTASK_TASK_BINARY_FILEPATH must first be set.  
TASKCLI_CFG_NO_GO_TASK  

### show_hidden_groups
Listing will show groups that were marked with hidden=True  
TASKCLI_CFG_SHOW_HIDDEN_GROUPS  

### show_hidden_tasks
Listing will show tasks that are marked with hidden=True  
TASKCLI_CFG_SHOW_HIDDEN_TASKS  

### show_tags
Listing will show tags of each task.  
TASKCLI_CFG_SHOW_TAGS  

### show_optional_args
Listing will show optional arguments of each task.  
TASKCLI_CFG_SHOW_OPTIONAL_ARGS  

### show_default_values
Listing will show default values of any arguments that are shown. Use with --show-optional-args to also show values of the optional arguments.  
TASKCLI_CFG_SHOW_DEFAULT_VALUES  

### print_task_start_message
Whether to print a log.info log message to stderr whenever a task starts.  
TASKCLI_CFG_PRINT_TASK_START_MESSAGE  

### show_ready_info
Listing tasks will show detailed info about the task's readiness to be run. For example, it will list any required but missing environment variables.   
TASKCLI_CFG_SHOW_READY_INFO  

### hide_not_ready
Tasks which are not ready to run (e.g. due to missing env vars) will be automatically marked as hidden.  
TASKCLI_CFG_HIDE_NOT_READY  

### print_env
List the supported env vars  
(no env var)  

### print_env_detailed
Like --print-env, but also include descriptions.  
(no env var)  

### verbose
Verbose output, show debug level logs.  
TASKCLI_CFG_VERBOSE  

### list
List tasks. Use -ll and -lll for a more detailed listing.  
(no env var)  

### print_return_value
Advanced: print return value of the task function to stdout. Useful when the task is a regular function which by itself does not print, and only returns a value.  
TASKCLI_CFG_PRINT_RETURN_VALUE  

### list_all
Listing tasks shows all possible infomation. Extremely very verbose output.  
TASKCLI_CFG_LIST_ALL  

### print_debug
Import the tasks and print detailed debug information. Use with, or without, specifying a task name.  
TASKCLI_CFG_PRINT_DEBUG  

### file
Which taskfiles to use by default, you can specify multiple (comma separated), they will be merged.default is 'tasks.py'  
TASKCLI_CFG_FILE  



docs/advanced_config.md
# Advanced Configuration
(This page was autogenerated, do not edit manually)
Summary of Advanced configuration options.
You typically Won't ever need to change these, but you can.
Most can be set either via a env var, or from within the tasks.py file.
If you use those, make sure to pin the version you use, as the advanced config
changes frequently.
## For fields containing format strings, here are the supported colors
none, dark_gray, white, yellow, green, pink, end, blue, red, underline, bold  

## List of field
### render_color_mandatory_arg (`str`)  
`default_value='\x1b[93m'`  
Env var: TASKCLI_ADV_RENDER_COLOR_MANDATORY_ARG  

### render_color_optional_arg (`str`)  
`default_value='\x1b[90m'`  
Env var: TASKCLI_ADV_RENDER_COLOR_OPTIONAL_ARG  

### render_color_default_arg (`str`)  
`default_value='\x1b[90m\x1b[4m'`  
Env var: TASKCLI_ADV_RENDER_COLOR_DEFAULT_ARG  

### render_color_summary (`str`)  
`default_value='\x1b[0m'`  
Env var: TASKCLI_ADV_RENDER_COLOR_SUMMARY  

### render_extra_line_indent (`str`)  
`default_value='  '`  
Env var: TASKCLI_ADV_RENDER_EXTRA_LINE_INDENT  

### render_format_regular_task_name (`str`)  
`default_value='{green}{name}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_REGULAR_TASK_NAME  

### render_format_important_tasks (`str`)  
`default_value='{green}{bold}{underline}{name}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_IMPORTANT_TASKS  

### render_format_hidden_tasks (`str`)  
`default_value='{dark_gray}{name}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_HIDDEN_TASKS  

### render_format_included_taskfile_dev_task (`str`)  
`default_value='{blue}{name}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_INCLUDED_TASKFILE_DEV_TASK  

### render_always_show_args (`list`)  
Always show these args in the task list, even if they are optional   
`default_value=[]`  
Env var: TASKCLI_ADV_RENDER_ALWAYS_SHOW_ARGS  

### render_highligh_params (`dict`)  
`default_value={}`  
Env var: TASKCLI_ADV_RENDER_HIGHLIGH_PARAMS  

### render_hide_optional_args (`list`)  
Never show optional arguments which have this name.  
Useful if many tasks have repeating, rarely used, optional arguments which we never.  
want to see in the task list.   
`default_value=[]`  
Env var: TASKCLI_ADV_RENDER_HIDE_OPTIONAL_ARGS  

### render_format_of_group_name (`str`)  
`default_value='{bold}{icon}{name:<22}{desc:<10}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_OF_GROUP_NAME  

### render_format_of_group_name_hidden (`str`)  
`default_value='{dark_gray}{bold}{name:<22}{desc:<10}{clear}'`  
Env var: TASKCLI_ADV_RENDER_FORMAT_OF_GROUP_NAME_HIDDEN  

### render_max_left_column_width (`int`)  
The left column (with task name and args) will pref  
Only rows with very long task names will be longer than this.  
If your task have many manadatory arguments, you may want to increase  
this to vertically align the summaries of tasks.   
`default_value=22`  
Env var: TASKCLI_ADV_RENDER_MAX_LEFT_COLUMN_WIDTH  

### render_min_left_column_width (`int`)  
`default_value=22`  
Env var: TASKCLI_ADV_RENDER_MIN_LEFT_COLUMN_WIDTH  

### render_max_params_per_line (`int`)  
Any tasks with more rendered params then this will be split into multiple lines   
`default_value=5`  
Env var: TASKCLI_ADV_RENDER_MAX_PARAMS_PER_LINE  

### render_max_default_arg_width (`int`)  
Default args longer than this will be truncated   
`default_value=40`  
Env var: TASKCLI_ADV_RENDER_MAX_DEFAULT_ARG_WIDTH  

### sort (`str`)  
Sort order   
`default_value='alpha'`  
Env var: TASKCLI_ADV_SORT  

### auto_hide_private_tasks (`bool`)  
`default_value=True`  
Env var: TASKCLI_ADV_AUTO_HIDE_PRIVATE_TASKS  

### name_namespace_separator (`str`)  
`default_value='.'`  
Env var: TASKCLI_ADV_NAME_NAMESPACE_SEPARATOR  

### alias_namespace_separator (`str`)  
`default_value=''`  
Env var: TASKCLI_ADV_ALIAS_NAMESPACE_SEPARATOR  



CHANGELOG.md
# 0.1.0
- Major rewrite of the internals.

# 0.0.3
- Moving to using `click` as the parsing backend.
- Moving to `click` changes the format of defining tasks.

# 0.0.2
- Initial release.


src/taskcli/configuration.py
import dataclasses
import sys

from . import envvars
from .timer import Timer

COLOR_NONE = ""
COLOR_DARK_GRAY = "\033[90m"
COLOR_WHITE = "\033[97m"
COLOR_YELLOW = "\033[93m"
COLOR_GREEN = "\033[92m"
COLOR_END = "\033[0m"
COLOR_UNDERLINE = "\033[4m"

@dataclasses.dataclass
class Colors:
    """Colors used for rendering the task list.

    Names of fields from this class can be used as '{color}' value in
    the RenderConfig object string format fields.
    """

    none: str = COLOR_NONE
    dark_gray: str = COLOR_DARK_GRAY
    white: str = COLOR_WHITE
    yellow: str = COLOR_YELLOW
    green: str = COLOR_GREEN
    pink: str = "\033[95m"
    end: str = COLOR_END
    blue: str = "\033[94m"
    red: str = "\033[91m"
    underline: str = COLOR_UNDERLINE
    bold: str = "\033[1m"



colors = Colors()

ENABLE_COLORS = sys.stdout.isatty()

if not ENABLE_COLORS:
    for key, _ in colors.__dict__.items():
        if key.startswith("__"):
            continue
        setattr(colors, key, "")

from typing import Any, Generic, TypeVar, Type

T = TypeVar("T")
import json

class _Field(Generic[T]):
    def __init__(self, value: T, /, desc: str = "", type_: Type|None = None, env: bool = True):
        self.value = value

        self.desc = desc.lstrip().rstrip()

        # Strip '        ' from the start of each line
        self.desc = "\n".join([line[8:] if line.startswith("        ") else line for line in self.desc.splitlines()])

        self.type = type_ if type_ else type(value)
        self.env = env

    def __get__(self, obj: Any, type: Any = None) -> T:
        return getattr(obj, '_' + self.name, self.value)

    def __set__(self, obj: Any, value: T) -> None:
        setattr(obj, '_' + self.name, value)

    def __set_name__(self, owner: Any, name: str) -> None:
        self.name = name

    def get_env_name(self, obj: Any) -> str:
        return "TASKCLI_ADV_" + self.name.upper()

    def cast(self, value: str) -> T:
        if self.type == list:
            return json.loads(value)
        elif self.type in [int, float, bool]:
            return self.type(value)

        raise Exception(f"Unhandled type {self.type}")
        return value  # Fallback for strings and unhandled types




class AdvConfig:
    """Advanced Config.

    You rarely need to change these settings, but you can if you e.g. want o create custom themes
    for your taskfiles.
    """

    # # The order of groups of tasks when list
    # # All tasks are by default in the "default" group unless task(group="foo") is used
    # # Any group not listed here will be shown last, in the order they were defined.

    render_color_mandatory_arg = _Field(colors.yellow)
    render_color_optional_arg = _Field(colors.dark_gray)
    render_color_optional_and_important_arg = ""
    render_color_default_arg = _Field(colors.dark_gray + colors.underline)
    render_color_summary = _Field(colors.end)
    render_extra_line_indent = _Field("  ")

    render_format_regular_task_name = _Field("{green}{name}{clear}")
    render_format_important_tasks = _Field("{green}{bold}{underline}{name}{clear}")
    render_format_hidden_tasks = _Field("{dark_gray}{name}{clear}")
    render_format_included_taskfile_dev_task = _Field("{blue}{name}{clear}")

    # Always show these args in the task list, even if they are optional
    render_always_show_args = _Field[list[str]]([], "Always show these args in the task list, even if they are optional")

    # Use to hide rarely used params,or params with long, names from the task list to reduce the noise
    # Hidden params still show up in the tab completion.

    # TODO hightlight soome arguments in special color
    render_highligh_params = _Field[dict[str,str]]({})


    render_hide_optional_args = _Field[list[str]]([], desc="""
        Never show optional arguments which have this name.
        Useful if many tasks have repeating, rarely used, optional arguments which we never.
        want to see in the task list.""")

    # use {NAME} instead of {name} to print group name in uppercase
    render_format_of_group_name = _Field("{bold}{icon}{name:<22}{desc:<10}{clear}")
    render_format_of_group_name_hidden = _Field("{dark_gray}{bold}{name:<22}{desc:<10}{clear}")


    render_max_left_column_width = _Field(22, desc="""
        The left column (with task name and args) will pref
        Only rows with very long task names will be longer than this.
        If your task have many manadatory arguments, you may want to increase
        this to vertically align the summaries of tasks.""")

    # The left column will never be smaller than this, even with short task names
    # Increaase it if you prefer your summaries to be more to the right.
    render_min_left_column_width = _Field(22)


    render_max_params_per_line = _Field(5, desc="""
        Any tasks with more rendered params then this will be split into multiple lines""")

    render_max_default_arg_width = _Field(40, desc="Default args longer than this will be truncated")

    sort = _Field("alpha", desc="Sort order")  # "alpha"  "definition")

    # If true, @task functions prefixed with "_" will be marked as hidden by default
    # No need to use hidden=True
    # e.g.
    #   @task
    #   def _mytask():
    #     ...
    auto_hide_private_tasks = _Field(True)

    # When set, it will be added
    name_namespace_separator = _Field(".")

    # When set, it will be added between the alias and the task name
    # By default it's empty, as it's often desired to keep aliases short and easy to type
    alias_namespace_separator = _Field("")


with Timer("Init Advanced config"):
    adv_config = AdvConfig()


def apply_simple_formatting() -> None:
    """Change the formatting to make the output simpler.

    This is useful for (and used by) unit tests to keep their assertions simpler.
    """
    adv_config.render_format_important_tasks = "{name}"
    adv_config.render_format_regular_task_name = "{name}"
    adv_config.render_format_hidden_tasks = "{name}"
    adv_config.render_format_included_taskfile_dev_task = "{name}"
    adv_config.render_format_of_group_name = "# {name}"
    adv_config.render_format_of_group_name_hidden = "# {name} HIDDEN"


if envvars.TASKCLI_ADV_OVERRIDE_FORMATTING.is_true():
    apply_simple_formatting()


def get_end_color() -> str:
    """Return the escape code to reset the terminal color."""
    if ENABLE_COLORS:
        return colors.end
    return ""


def get_underline() -> str:
    """Return the escape code to underline the text."""
    if ENABLE_COLORS:
        return COLOR_UNDERLINE
    return ""
